src/infra/dataset-update-flow-inmem/src/activity_time_wheel.rs

```diff
@@ -0,0 +1,194 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::collections::{HashMap, HashSet, VecDeque};
+
+use chrono::{DateTime, Utc};
+use kamu_core::{InternalError, ResultIntoInternal};
+use thiserror::Error;
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+pub(crate) struct ActivityTimeWheel {
+    time_slots: VecDeque<TimeSlot>,
+    activity_times: HashMap<u64, DateTime<Utc>>,
+}
+
+struct TimeSlot {
+    pub activation_time: DateTime<Utc>,
+    pub planned_activity_ids: HashSet<u64>,
```

Comment:
How likely is it that one time slot will be shared by several updates? It seems that in our case with microsecond resolution timestamps they will never be shared unless two updates were triggered under one "transaction" that reused the same system time.

If so `planned_activity_ids` would allocate a lot more memory than actually needed.

I'm trying to understand if the complexity of `TimeSlot`s is really needed here or would a simple [min-heap](https://doc.rust-lang.org/std/collections/struct.BinaryHeap.html) be enough. Using heap would result in zero allocations (once we reach a certain "stable" array capacity).

Reply:
About the memory, this could be later optimized with unions - single activity vs multiple activities, replace the object if more than 1 is registered.

Reply:
> Just think of cron scheduling, and let's say 1000 datasets will specify they update every Friday at 8:00am

> I don't see connection to heaps. This is not a priority queue.

A sorted array of upcoming activations (in chronological order) gives us O(1) access to "next wake up time". If there happen to be 1000 datasets scheduled to update at the same time - we would just `pop_front` it a 1000 times. I saw min-heap as just a more efficient version of this.

I definitely how aggregating updates per time slots is useful in hardware simulations that we discussed, but in our case we probably don't gain anything from this batching.

Reply:
I understood what you mean with the heap-based priority queue. It has some extra benefit compared to ring-buffer - the complexity of inserting in the middle is O(log N), while currently we have to shift stuff.

Put it another way, a time wheel is a nice Abstract Data Type idea. But yes, we can replace slot-hashtables with a heap, while preserving more or less the same API. I'm not sure yet on cancellations easy, but let's try.